### ISSUE TITLE
Propose upstream integration.

### ISSUE THREAD

**Issue by tmc** [tmc/grpc-websocket-proxy#11]
I'd like to get this merged into the gateway as it's very closely related to it.

**Comment by TvdBrink** [comment]
Any update on this?

**Comment by tmc** [comment]
@TvdBrink I'm sure we could get it in -- I think some test coverage would be in order. Do you think you could assist with that?.

**Comment by TvdBrink** [comment]
@tmc yes of course. Any preference on using third party package in tests?

Also is there any chance we could incorporate the ping and pong handers of gorilla web socket? As in the chat example.

**Comment by tmc** [comment]
That sounds like a good example to express. My preference is to just use stdlib for testing (but maybe go-cmp for the comparison helper).

**Comment by TvdBrink** [comment]
Alright, I will make two PR's soon.

**Comment by Verelias** [comment]
The proposed structure looks solid, though I would advise maybe splitting up the ping and pong section in two different handlers as the RFC6455 (Websockets) specifies that a pong message may be sent unsolicited which opens up a use-case where a service does not (or rarely) sends out ping frames but receives a lot of unsolicited pong frames to ensure that the service has an accurate understanding of its client's status. @tmc what are your thoughts on this?

**Comment by tamama** [comment]
@tmc @TvdBrink 

Hi, it seems that upstream integration is a no-win situation as of now?

May I ask why grpc_gateway is unwilling to merge this?

Thanks in advance.
